File: Client/Protocol/Response.cpp
//
//  Created by Stasel
//

#include "Response.h"

// MARK: - Response
Response::Response(UInt8 version, ResponseCode code) : version(version), code(code) {}
UInt8 Response::getVersion() const { return version; }
ResponseCode Response::getCode() const { return code; }


// MARK: - RegisterSuccessResponse
RegisterSuccessResponse::RegisterSuccessResponse(UInt8 version, const UUID& clientID)
    : Response(version, RESPONSE_REGISTER_SUCCESS), clientID(clientID) {}

const UUID& RegisterSuccessResponse::getClientID() const {
    return clientID;
}

RegisterSuccessResponse* RegisterSuccessResponse::fromPayload(UInt8 version, const Bytes& payload) {
    if (payload.size() != UUID_SIZE) {
        throw std::runtime_error("Invalid payload size");
    }
    return new RegisterSuccessResponse(version, UUID(payload));
}


// MARK: - ClientListResponse
ClientListResponse::ClientListResponse(UInt8 version, const std::vector<RemoteClient>& clients)
    : Response(version, RESPONSE_CLIENT_LIST), clients(clients) { }
const std::vector<RemoteClient>& ClientListResponse::getClients() const {
    return clients;
}

ClientListResponse* ClientListResponse::fromPayload(UInt8 version, const Bytes& payload) {
    if (payload.size() % (UUID_SIZE + USERNAME_MAX_LENGTH) != 0) {
        throw std::runtime_error("Invalid payload size");
    }
    std::vector<RemoteClient> clientList;
    var it = payload.begin();
    while (it != payload.end()) {
        const Bytes clientID(it, it + UUID_SIZE);
        it += UUID_SIZE;
        String clientName(it, it+ USERNAME_MAX_LENGTH);
        
        // Remove trailing null characters:
        clientName.erase(std::find(clientName.begin(), clientName.end(), '\0'), clientName.end());
        
        it += USERNAME_MAX_LENGTH;
        clientList.push_back(RemoteClient(UUID(clientID), clientName));
    }
    return new ClientListResponse(version, clientList);
}


// MARK: - PublicKeyResponse
PublicKeyResponse::PublicKeyResponse(UInt8 version, const UUID& clientID, const Bytes& publicKey)
    : Response(version, RESPONSE_PUBLIC_KEY), clientID(clientID), publicKey(publicKey) {}
const Bytes& PublicKeyResponse::getPublicKey() const {
    return publicKey;
}

const UUID& PublicKeyResponse::getClientID() const {
    return clientID;
}

PublicKeyResponse* PublicKeyResponse::fromPayload(UInt8 version, const Bytes& payload) {
    if (payload.size() < UUID_SIZE) {
        throw std::runtime_error("Invalid payload size");
    }
    const Bytes clientID(payload.begin(), payload.begin() + UUID_SIZE);
    const Bytes publicKey(payload.begin() + UUID_SIZE, payload.end());
    return new PublicKeyResponse(version, UUID(clientID), publicKey);
}


// MARK: - MessageSentResponse
MessageSentResponse::MessageSentResponse(UInt8 version, const UUID& toClient, UInt32 messageID)
    : Response(version, RESPONSE_MESSAGE_SENT), toClient(toClient), messageID(messageID) { }
const UUID& MessageSentResponse::getToClient() const {
    return toClient;
}

const UInt32 MessageSentResponse::getMessageID() const {
    return messageID;
}

MessageSentResponse* MessageSentResponse::fromPayload(UInt8 version, const Bytes& payload) {
    if (payload.size() != UUID_SIZE + MESSAGE_ID_SIZE) {
        throw std::runtime_error("Invalid payload size");
    }
    const Bytes clientID(payload.begin(), payload.begin() + UUID_SIZE);
    const Bytes messageIDSplitted(payload.begin() + UUID_SIZE, payload.end());
    let messageID = joinBytes32((UInt8*)messageIDSplitted.data());
    return new MessageSentResponse(version, UUID(clientID), messageID);
}


// MARK: - MessageListResponse
MessageListResponse::MessageListResponse(UInt8 version, const std::vector<Message>& messages)
    : Response(version, RESPONSE_MESSAGE_LIST), messages(messages) { }

const std::vector<Message>& MessageListResponse::getMessages() const {
    return messages;
}

MessageListResponse* MessageListResponse::fromPayload(UInt8 version, const Bytes& payload) {
    std::vector<Message> messages;
    var it = payload.begin();
    while (it != payload.end()) {
        const Bytes senderClientID(it, it + UUID_SIZE);
        it += UUID_SIZE;
        let messageID = joinBytes32((UInt8*)Bytes(it, it + MESSAGE_ID_SIZE).data());
        it += MESSAGE_ID_SIZE;
        let messageType = *it;
        it++;
        let messageSize = joinBytes32((UInt8*)Bytes(it, it + 4).data());
        it += 4;
        Bytes messageContent(it, it + messageSize);
        it += messageSize;
        messages.push_back(Message(UUID(senderClientID), messageID, (MessageType)messageType, messageContent));
    }
    return new MessageListResponse(version, messages);
}


// MARK: - ServerErrorResponse
ServerErrorResponse::ServerErrorResponse(UInt8 version): Response(version, RESPONSE_SERVER_ERROR) {}

File: Server/Protocol/Response.py
from __future__ import annotations
import uuid
from enum import Enum
from Message import Message
from Client import Client
import struct
import Constants


class ResponseCode(Enum):
    REGISTER_SUCCESS = 2000
    CLIENT_LIST = 2001
    PUBLIC_KEY = 2002
    MESSAGE_SENT = 2003
    MESSAGE_LIST = 2004
    SERVER_ERROR = 9000


class Response:
    def __init__(self, version: int, code: ResponseCode):
        assert isinstance(version, int)
        assert isinstance(code, ResponseCode)
        self.version = version
        self.code = code

    def pack(self, payload=bytes()) -> bytes:
        return struct.pack("<BHI", self.version, self.code.value, len(payload)) + payload

    def __str__(self) -> str:
        return f"[Response] {self.code.value}({self.code.name})"


class RegisterSuccessResponse(Response):
    def __init__(self, version: int, clientID: uuid.UUID):
        super().__init__(version, ResponseCode.REGISTER_SUCCESS)
        assert isinstance(clientID, uuid.UUID)
        self.clientID = clientID

    def pack(self) -> bytes:
        payload = self.clientID.bytes
        return super().pack(payload)


class ClientListResponse(Response):
    def __init__(self, version: int, clientList: [Client]):
        super().__init__(version, ResponseCode.CLIENT_LIST)
        assert isinstance(clientList, list)
        self.clientList = clientList

    def pack(self) -> bytes:
        payload = bytes()
        for client in self.clientList:
            payload += client.id.bytes
            payload += bytes(client.name, Constants.STR_ENCODING).ljust(Constants.CLIENT_NAME_MAX_LENGTH, b'\0')
        return super().pack(payload)


class PublicKeyResponse(Response):
    def __init__(self, version: int, clientID: uuid.UUID, publicKey: bytes):
        super().__init__(version, ResponseCode.PUBLIC_KEY)
        assert isinstance(clientID, uuid.UUID)
        assert isinstance(publicKey, bytes)
        self.clientID = clientID
        self.publicKey = publicKey

    def pack(self) -> bytes:
        payload = self.clientID.bytes + self.publicKey
        return super().pack(payload)


class MessageSentResponse(Response):
    def __init__(self, version: int, toClientID: uuid.UUID, messageID: int):
        super().__init__(version, ResponseCode.MESSAGE_SENT)
        assert isinstance(toClientID, uuid.UUID)
        assert isinstance(messageID, int)
        self.toClientID = toClientID
        self.messageID = messageID

    def pack(self) -> bytes:
        payload = self.toClientID.bytes + struct.pack("<I", self.messageID)
        return super().pack(payload)


class MessageListResponse(Response):
    def __init__(self, version: int, messages: [Message]):
        super().__init__(version, ResponseCode.MESSAGE_LIST)
        assert isinstance(messages, list)
        self.messages = messages

    def pack(self) -> bytes:
        payload = bytes()
        for message in self.messages:
            payload += message.fromClient.bytes
            payload += struct.pack("<IBI", message.id, message.type, len(message.content))
            payload += message.content
        return super().pack(payload)


class ServerErrorResponse(Response):
    def __init__(self, version: int):
        super().__init__(version, ResponseCode.SERVER_ERROR)
